### ISSUE TITLE
Iterator adapter like std::iter::Inspect, but which passes a mutable reference to the closure

### ISSUE THREAD

**Issue by abonander** [bluss/rust-itertools#82]
I don't know how common of a use-case this is, but I can see some utility in an iterator adapter which can mutate items via a closure before yielding them. I was going to open this as an issue on rust-lang/rfcs but I figured it would be better to incubate the idea here first.

Basically a more succinct version of this:

``` rust
my_iter.map(|mut val| {
    val.mutate();
    val
})
```

As for what to call it, I'm not sure. Perhaps as a parallel to `.inspect()` but implying the possibility of change. Here's some of my ideas so far:
- `.mutate()`
- `.edit()` (as a reference to `sed`)
- `.update()` 
- `.probe()`
- `.modify()`
- `.tweak()`

**Comment by bluss** [comment]
Maybe the name `.map_inplace()`? Could even be a win for regular mapping in some weird scenario where moves are costly and don't optimize out.

Do you have a real use case in mind? Asking because we don't want to collect nice to haves that nobody needed.

**Comment by cybergeek94** [comment]
One example came up in a possible improvement to a solution I gave in [this thread on /r/rust](https://www.reddit.com/r/rust/comments/3xscgi/help_with_code/cy7kbe3) (with your suggested naming):

```rust

fn permutations(seq: Vec<i32>) -> Vec<Vec<i32>> {
    match seq.len() {
        0 => return vec![],
        1 => return vec![seq],
        _ => (),
    }

    seq.iter().enumerate().flat_map(|(head_idx, &head)| {
        let mut remainder = seq[..head_idx].to_owned();

        if seq.len() > head_idx + 1 {
            remainder.extend(&seq[head_idx + 1..]);
        }

        permutations(remainder).map_inplace(|perm| perm.insert(0, head))
    })
    .collect()
}
```

**Comment by bluss** [comment]
I like update too.

**Comment by bluss** [comment]
Implemented by @lucasem in #232 and in itertools 0.7.4!
